Panic in mylogical/conn.go; index out of bounds

https://github.com/cockroachdb/replicator/actions/runs/9024225261/job/24797705112#step:7:6052

```
=== RUN   TestMYLogical/consistent
time="2024-05-09T22:18:28Z" level=info msg="source connect string: postgresql://root@localhost:26257/defaultdb?sslmode=disable"
time="2024-05-09T22:18:28Z" level=info msg="source schema: \"src-13173-1\".\"public\""
time="2024-05-09T22:18:28Z" level=info msg="staging connect string: postgresql://root@localhost:26257/defaultdb?sslmode=disable"
time="2024-05-09T22:18:29Z" level=info msg="reusing SourcePool as TargetPool"
time="2024-05-09T22:18:29Z" level=info msg="target schema: \"tgt-13173-3\".\"public\""
time="2024-05-09T22:18:29Z" level=info msg="Pool: Setup 1 new connections (minimal pool size)..." mysql=true
time="2024-05-09T22:18:29Z" level=info msg="Version info: 5.7.44-log"
time="2024-05-09T22:18:29Z" level=warning msg="Detecting MySQL 5.X; forcing metadata fetch"
time="2024-05-09T22:18:29Z" level=info msg="Master status: ea865bc2-0e51-11ef-aced-0242ac120002:1-8"
time="2024-05-09T22:18:29Z" level=info msg="gtidSet: ea865bc2-0e51-11ef-aced-0242ac120002:1-8"
time="2024-05-09T22:18:30Z" level=info msg="Version info: 5.7.44-log"
time="2024-05-09T22:18:30Z" level=warning msg="Detecting MySQL 5.X; forcing metadata fetch"
[2024/05/09 22:18:30] [info] binlogsyncer.go:156 create BinlogSyncer with config {123456 mysql localhost 3306 root    false false <nil> false UTC false 0 0s 0s 0 false false 0 <nil> 0xc000488900}
[2024/05/09 22:18:30] [info] binlogsyncer.go:410 begin to sync binlog from GTID set ea865bc2-0e51-11ef-aced-0242ac120002:1-8
[2024/05/09 22:18:30] [info] binlogsyncer.go:808 rotate to (log-bin.000003, 4)
time="2024-05-09T22:18:30Z" level=info msg="connected to MySQL version 5.7.44-log"
[2024/05/09 22:18:32] [info] binlogsyncer.go:188 syncer is closing...
[2024/05/09 22:18:32] [info] binlogsyncer.go:881 kill last connection id 7
[2024/05/09 22:18:32] [info] binlogsyncer.go:218 syncer is closed
time="2024-05-09T22:18:32Z" level=info msg="finished my pool cleanup"
=== RUN   TestMYLogical/consistent-chaos
time="2024-05-09T22:18:32Z" level=info msg="source connect string: postgresql://root@localhost:26257/defaultdb?sslmode=disable"
time="2024-05-09T22:18:32Z" level=info msg="source schema: \"src-13173-4\".\"public\""
time="2024-05-09T22:18:32Z" level=info msg="staging connect string: postgresql://root@localhost:26257/defaultdb?sslmode=disable"
time="2024-05-09T22:18:32Z" level=info msg="reusing SourcePool as TargetPool"
time="2024-05-09T22:18:32Z" level=info msg="target schema: \"tgt-13173-6\".\"public\""
time="2024-05-09T22:18:33Z" level=info msg="Pool: Setup 1 new connections (minimal pool size)..." mysql=true
time="2024-05-09T22:18:33Z" level=info msg="Version info: 5.7.44-log"
time="2024-05-09T22:18:33Z" level=warning msg="Detecting MySQL 5.X; forcing metadata fetch"
time="2024-05-09T22:18:33Z" level=info msg="Master status: ea865bc2-0e51-11ef-aced-0242ac120002:1-14"
time="2024-05-09T22:18:33Z" level=info msg="gtidSet: ea865bc2-0e51-11ef-aced-0242ac120002:1-14"
time="2024-05-09T22:18:33Z" level=info msg="Version info: 5.7.44-log"
time="2024-05-09T22:18:33Z" level=warning msg="Detecting MySQL 5.X; forcing metadata fetch"
[2024/05/09 22:18:33] [info] binlogsyncer.go:156 create BinlogSyncer with config {123456 mysql localhost 3306 root    false false <nil> false UTC false 0 0s 0s 0 false false 0 <nil> 0xc000196420}
[2024/05/09 22:18:33] [info] binlogsyncer.go:410 begin to sync binlog from GTID set ea865bc2-0e51-11ef-aced-0242ac120002:1-14
[2024/05/09 22:18:33] [info] binlogsyncer.go:808 rotate to (log-bin.000003, 4)
time="2024-05-09T22:18:33Z" level=info msg="connected to MySQL version 5.7.44-log"
[2024/05/09 22:18:34] [info] binlogsyncer.go:188 syncer is closing...
[2024/05/09 22:18:34] [info] binlogsyncer.go:881 kill last connection id 15
[2024/05/09 22:18:34] [info] binlogsyncer.go:218 syncer is closed
time="2024-05-09T22:18:34Z" level=warning msg="error while copying messages; will retry" error=chaos
[2024/05/09 22:18:35] [info] binlogsyncer.go:156 create BinlogSyncer with config {123456 mysql localhost 3306 root    false false <nil> false UTC false 0 0s 0s 0 false false 0 <nil> 0xc000c68600}
[2024/05/09 22:18:35] [info] binlogsyncer.go:410 begin to sync binlog from GTID set ea865bc2-0e51-11ef-aced-0242ac120002:15
[2024/05/09 22:18:35] [info] binlogsyncer.go:808 rotate to (log-bin.000002, 4)
time="2024-05-09T22:18:35Z" level=info msg="connected to MySQL version 5.7.44-log"
[2024/05/09 22:18:35] [info] binlogsyncer.go:188 syncer is closing...
[2024/05/09 22:18:35] [info] binlogsyncer.go:881 kill last connection id 18
[2024/05/09 22:18:35] [info] binlogsyncer.go:218 syncer is closed
panic: runtime error: index out of range [2] with length 2

goroutine 2704 [running]:
github.com/cockroachdb/replicator/internal/source/mylogical.(*conn).onDataTuple(0xc0001a6180, {0xc0000cc540?, 0x1d83771?}, 0xc000c80d60, 0xc0005a15f0, 0x1)
	/home/runner/work/replicator/replicator/internal/source/mylogical/conn.go:365 +0x17ed
github.com/cockroachdb/replicator/internal/source/mylogical.(*conn).accumulateBatch(0xc0001a6180, 0xc000348900, 0xc000ad71d0, 0xc000c80d60)
	/home/runner/work/replicator/replicator/internal/source/mylogical/conn.go:247 +0x1213
github.com/cockroachdb/replicator/internal/source/mylogical.(*conn).copyMessages(0xc0001a6180, 0xc000348900)
	/home/runner/work/replicator/replicator/internal/source/mylogical/conn.go:303 +0xc05
github.com/cockroachdb/replicator/internal/source/mylogical.(*conn).Start.func1()
	/home/runner/work/replicator/replicator/internal/source/mylogical/conn.go:121 +0x7c
github.com/cockroachdb/replicator/internal/util/stopper.(*Context).Go.func1()
	/home/runner/work/replicator/replicator/internal/util/stopper/stopper.go:187 +0xbe
created by github.com/cockroachdb/replicator/internal/util/stopper.(*Context).Go in goroutine 2590
	/home/runner/work/replicator/replicator/internal/util/stopper/stopper.go:185 +0x14d
FAIL	github.com/cockroachdb/replicator/internal/source/mylogical	7.255s
time="2024-05-09T22:18:32Z" level=info msg="skipping postgresql integration tests: CDC_INTEGRATION=\"mysql-v5.7\""
ok  	github.com/cockroachdb/replicator/internal/source/pglogical	1.070s
	github.com/cockroachdb/replicator/internal/util/stdserver		coverage: 0.0% of statements
	github.com/cockroachdb/replicator/internal/util/stmtcache		coverage: 0.0% of statements
	github.com/cockroachdb/replicator/internal/util/subfs		coverage: 0.0% of statements
```

--- a/internal/source/mylogical/conn.go
+++ b/internal/source/mylogical/conn.go
@@ -244,7 +244,7 @@ func (c *conn) accumulateBatch(
 			return nil, errors.Errorf("Operation not supported %s", ev.Header.EventType)
 		}
 		mutationCount.With(prometheus.Labels{"type": operation.String()}).Inc()
-		if err := c.onDataTuple(ctx, batch, e, operation); err != nil {
+		if err := c.onDataTuple(batch, e, operation); err != nil {
 			return nil, err
 		}
 
@@ -333,10 +333,7 @@ func (c *conn) ZeroStamp() stamp.Stamp {
 }
 
 func (c *conn) onDataTuple(
-	ctx context.Context,
-	batch *types.MultiBatch,
-	tuple *replication.RowsEvent,
-	operation mutationType,
+	batch types.Accumulator, tuple *replication.RowsEvent, operation mutationType,
 ) error {
 	tbl, ok := c.relations[tuple.TableID]
 	if !ok {
@@ -352,15 +349,18 @@ func (c *conn) onDataTuple(
 	}
 	log.Tracef("%s on table %s (#rows: %d)", operation, tbl, len(tuple.Rows))
 	for rowNum, row := range tuple.Rows {
-		var err error
-		var key []any
-		var mut types.Mutation
-		enc := make(map[string]any)
 		// on update we only care about the new value.
 		// even rows are skipped since they contain the value before the update
 		if operation == updateMutation && (rowNum%2 == 0) {
 			continue
 		}
+		if len(targetCols) != len(row) {
+			return errors.Errorf("unexpected number of columns in the logical stream for %s", tbl)
+		}
+		var err error
+		var key []any
+		var mut types.Mutation
+		enc := make(map[string]any)
 		for idx, sourceCol := range row {
 			targetCol := targetCols[idx]
 			switch s := sourceCol.(type) {
@@ -383,7 +383,7 @@ func (c *conn) onDataTuple(
 			}
 		}
 		if len(key) == 0 && operation != insertMutation {
-			return errors.Errorf("only inserts supported with no key")
+			return errors.Errorf("only inserts supported with no key for %s", tbl)
 		}
 		mut.Key, err = json.Marshal(key)
 		if err != nil {

--- a/internal/source/mylogical/conn_test.go
+++ b/internal/source/mylogical/conn_test.go
@@ -1,0 +1,232 @@
+// Copyright 2024 The Cockroach Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package mylogical
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/cockroachdb/replicator/internal/types"
+	"github.com/cockroachdb/replicator/internal/util/hlc"
+	"github.com/cockroachdb/replicator/internal/util/ident"
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/replication"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+type mockAccumulator struct {
+	mutations ident.Map[[]types.Mutation]
+}
+
+var _ types.Accumulator = &mockAccumulator{}
+
+// Accumulate implements Accumulator.
+func (m *mockAccumulator) Accumulate(table ident.Table, mut types.Mutation) error {
+	mutations, ok := m.mutations.Get(table.Table())
+	if !ok {
+		mutations = make([]types.Mutation, 0)
+	}
+	mutations = append(mutations, mut)
+	m.mutations.Put(table.Table(), mutations)
+	return nil
+}
+
+func (m *mockAccumulator) compare(a *assert.Assertions, table ident.Table, muts []types.Mutation) {
+	mutations, ok := m.mutations.Get(table.Table())
+	if !ok {
+		a.Failf("unknown table %s", table.Raw())
+		return
+	}
+	if len(mutations) != len(muts) {
+		a.Fail("mutations are not the same")
+		return
+	}
+	for idx, mut := range mutations {
+		a.Equal(muts[idx].Before, mut.Before)
+		a.Equal(muts[idx].Data, mut.Data)
+		a.Equal(muts[idx].Key, mut.Key)
+		a.Equal(muts[idx].Time, mut.Time)
+		a.NotNil(mut.Meta["mylogical"])
+	}
+}
+
+// TestOnDataTuple verifies that incoming tuples are added to a batch
+func TestOnDataTuple(t *testing.T) {
+	r := require.New(t)
+	consistentPoint, err := newConsistentPoint(mysql.MariaDBFlavor).parseFrom("1-1-1")
+	ts := hlc.New(consistentPoint.AsTime().UnixNano(), 0)
+	r.NoError(err)
+	schema := ident.MustSchema(ident.Public)
+	// Simple KV table
+	kvTableID := uint64(0)
+	kvTable := ident.NewTable(schema, ident.New("t1"))
+	kvCols := []types.ColData{
+		{Name: ident.New("k"), Primary: true, Type: "int"},
+		{Name: ident.New("v"), Primary: false, Type: "int"},
+	}
+	// Table with no key
+	noKeyID := uint64(1)
+	noKeyTable := ident.NewTable(schema, ident.New("t2"))
+	noKeyCols := []types.ColData{
+		{Name: ident.New("v"), Primary: false, Type: "int"},
+	}
+	tables := []ident.Table{kvTable, noKeyTable}
+	columns := &ident.TableMap[[]types.ColData]{}
+	columns.Put(kvTable, kvCols)
+	columns.Put(noKeyTable, noKeyCols)
+	c := &conn{
+		columns:             columns,
+		nextConsistentPoint: consistentPoint,
+		relations: map[uint64]ident.Table{
+			kvTableID: kvTable,
+			noKeyID:   noKeyTable,
+		},
+		target: schema,
+	}
+	tests := []struct {
+		name      string
+		tuple     *replication.RowsEvent
+		operation mutationType
+		wantMuts  []types.Mutation
+		wantErr   string
+	}{
+		{
+			name: "insert",
+			tuple: &replication.RowsEvent{
+				TableID: kvTableID,
+				Rows: [][]any{
+					{1, 10},
+					{2, 20},
+					{3, 30},
+				},
+			},
+			operation: insertMutation,
+			wantMuts: []types.Mutation{
+				{
+					Data: json.RawMessage(`{"k":1,"v":10}`),
+					Key:  json.RawMessage(`[1]`),
+					Time: ts,
+				},
+				{
+					Data: json.RawMessage(`{"k":2,"v":20}`),
+					Key:  json.RawMessage(`[2]`),
+					Time: ts,
+				},
+				{
+					Data: json.RawMessage(`{"k":3,"v":30}`),
+					Key:  json.RawMessage(`[3]`),
+					Time: ts,
+				},
+			},
+		},
+		{
+			name: "insert no key",
+			tuple: &replication.RowsEvent{
+				TableID: noKeyID,
+				Rows: [][]any{
+					{1},
+				},
+			},
+			operation: insertMutation,
+			wantMuts: []types.Mutation{
+				{
+					Data: json.RawMessage(`{"v":1}`),
+					Key:  json.RawMessage(`null`),
+					Time: ts,
+				},
+			},
+		},
+		{
+			name: "update",
+			tuple: &replication.RowsEvent{
+				TableID: kvTableID,
+				Rows: [][]any{
+					{1, 10},
+					{1, 11},
+					{2, 20},
+					{2, 21},
+				},
+			},
+			operation: updateMutation,
+			wantMuts: []types.Mutation{
+				{
+					Data: json.RawMessage(`{"k":1,"v":11}`),
+					Key:  json.RawMessage(`[1]`),
+					Time: ts,
+				},
+				{
+					Data: json.RawMessage(`{"k":2,"v":21}`),
+					Key:  json.RawMessage(`[2]`),
+					Time: ts,
+				},
+			},
+		},
+		{
+			name: "update no key",
+			tuple: &replication.RowsEvent{
+				TableID: noKeyID,
+				Rows: [][]any{
+					{1},
+					{1},
+				},
+			},
+			operation: updateMutation,
+			wantErr:   "only inserts supported with no key",
+		},
+		{
+			name: "delete",
+			tuple: &replication.RowsEvent{
+				TableID: kvTableID,
+				Rows: [][]any{
+					{3, 2},
+				},
+			},
+			operation: deleteMutation,
+			wantMuts: []types.Mutation{
+				{
+					Key:  json.RawMessage(`[3]`),
+					Time: ts,
+				},
+			},
+		},
+		{
+			name: "invalid_row_size", // Verification for bug #858
+			tuple: &replication.RowsEvent{
+				TableID: kvTableID,
+				Rows: [][]any{
+					{1, 2, 3},
+				},
+			},
+			operation: insertMutation,
+			wantErr:   "unexpected number of columns",
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			a := assert.New(t)
+			batch := &mockAccumulator{}
+			err := c.onDataTuple(batch, tt.tuple, tt.operation)
+			if tt.wantErr != "" {
+				a.ErrorContains(err, tt.wantErr)
+				return
+			}
+			a.NoError(err)
+			batch.compare(a, tables[tt.tuple.TableID], tt.wantMuts)
+		})
+	}
+}

